tests/Spec.hs

```diff
@@ -145,142 +149,160 @@ defaultResults = Results
+        Writer.tell [AGetOpenPullRequests]
+        takeResultGetOpenPullRequests
+      GetLatestVersion _ -> takeResultGetLatestVersion
+      GetChangelog _ _ -> takeResultGetChangelog
+      IncreaseMergeMetric -> pure ()
+      UpdateTrainSizeMetric n -> do
+        results <- State.get
+        State.put $ results { resultTrainSizeUpdates = n : resultTrainSizeUpdates results }
+        pure ()
+
+runActionEff
+  :: (State Results :> es, Writer [ActionFlat] :> es)
+  => Config.ProjectConfiguration
+  -> Eff (BaseAction : RetrieveEnvironment : es) a
+  -> Eff es a
+runActionEff config eff = runRetrieveInfo config $ runBaseActionResults eff
 
 -- Simulates running the action. Use the provided results as result for various
 -- operations. Results are consumed one by one.
-runActionCustom :: HasCallStack => Results -> Action a -> (a, [ActionFlat])
-runActionCustom results action = Rws.evalRWS (runActionRws testProjectConfig action) () results
 
-runActionCustomResults :: HasCallStack => Results -> Action a -> (a, Results, [ActionFlat])
-runActionCustomResults results action = Rws.runRWS (runActionRws testProjectConfig action) () results
+runActionCustom :: Results -> Eff ActionResults a -> (a, [ActionFlat])
+runActionCustom results action = runPureEff $ Writer.runWriter $ State.evalState results $ runActionEff testProjectConfig action
+
+runActionCustomResults :: Results -> Eff ActionResults a -> (a, Results, [ActionFlat])
+runActionCustomResults results action =
+    flatten $ runPureEff $ Writer.runWriter $ State.runState results $ runActionEff testProjectConfig action
+  where
+    flatten ((a, b), c) = (a, b, c)
 
-runActionCustomConfig :: HasCallStack => Config.ProjectConfiguration -> Results -> Action a -> (a, [ActionFlat])
-runActionCustomConfig projectConfig results action = Rws.evalRWS (runActionRws projectConfig action) () results
+runActionCustomConfig :: Config.ProjectConfiguration -> Results -> Eff ActionResults a -> (a, [ActionFlat])
+runActionCustomConfig projectConfig results action =
+  runPureEff $ Writer.runWriter $ State.evalState results $ runActionEff projectConfig action
 
 -- Simulates running the action with default results.
-runAction :: Action a -> (a, [ActionFlat])
+runAction :: Eff ActionResults a -> (a, [ActionFlat])
 runAction = runActionCustom defaultResults
 
 -- Handle an event, then advance the state until a fixed point,
 -- and simulate its side effects.
-handleEventTest :: Event -> ProjectState -> Action ProjectState
+handleEventTest :: (BaseAction :> es, RetrieveEnvironment :> es) => Event -> ProjectState -> Eff es ProjectState
 handleEventTest = Logic.handleEvent testTriggerConfig testmergeWindowExemptionConfig
 
 -- Handle events (advancing the state until a fixed point in between) and
 -- simulate their side effects.
-handleEventsTest :: [Event] -> ProjectState -> Action ProjectState
+handleEventsTest :: (BaseAction :> es, RetrieveEnvironment :> es) => [Event] -> ProjectState -> Eff es ProjectState
```

Comment:
One funny thing to finally note is that the tests themselves are not changed at all. All the changes are limited to the boilerplate. It really goes to show that free monads and effect systems serve the exact same purpose, be it with different jargon and ways of going at it.